lib/THC/THCReduce.cuh

```diff
@@ -0,0 +1,319 @@
+                                ModifyOp modifyOp,
+                                ReduceOp reduceOp) {
+  const IndexType sliceIndex = getReduceNoncontigDimSliceIndex<IndexType>();
+
+  if (sliceIndex >= totalSlices) {
+    return;
+  }
+
+  // Each thread picks a point in `out` and `in` for which it is
+  // producing the reduction
+  const IndexType outOffset =
+    IndexToOffset<IndexType, ADims>::get(sliceIndex, out);
+  const IndexType inBaseOffset =
+    IndexToOffset<IndexType, BDims>::get(sliceIndex, in);
+
+  // For each point in reductionSize, reduce into `r`
+  IndexType inOffset = inBaseOffset;
+  float r = init;
+
+  for (IndexType i = 0; i < reductionSize; ++i) {
+    r = reduceOp(r, modifyOp(in.data[inOffset]));
+    inOffset += reductionStride;
+  }
+
+  // Write out reduced value
+  out.data[outOffset] = r;
+}
+
+template <typename IndexType>
+__device__ __forceinline__ IndexType getReduceContigDimSliceIndex() {
+  // Each block handles one slice
+  return getLinearBlockId<IndexType>();
+}
+
+// Kernel that handles an entire reduction of a slice of a tensor per
+// each block
+template <typename ModifyOp, typename ReduceOp, typename IndexType, int ADims, int BDims>
+__global__ void
+THCudaTensor_reduceContigDim(TensorInfo<IndexType> out,
+                             TensorInfo<IndexType> in,
+                             IndexType reductionSize,
+                             IndexType totalSlices,
+                             float init,
+                             ModifyOp modifyOp,
+                             ReduceOp reduceOp) {
+  const IndexType sliceIndex = getReduceContigDimSliceIndex<IndexType>();
+
+  if (sliceIndex >= totalSlices) {
+    return;
+  }
+
+  // Get the offset in `out` for the reduction
+  const IndexType outOffset =
+    IndexToOffset<IndexType, ADims>::get(sliceIndex, out);
+
+  // Get the base offset in `in` for this block's reduction
+  const IndexType inBaseOffset =
+    IndexToOffset<IndexType, BDims>::get(sliceIndex, in);
+
+  // Each thread in the block will reduce some subset of elements in
+  // the slice. The elements are guaranteed contiguous starting at
+  // `inBaseOffset`.
+  float r = init;
+  for (IndexType i = threadIdx.x; i < reductionSize; i += blockDim.x) {
+    r = reduceOp(r, modifyOp(in.data[inBaseOffset + i]));
+  }
+
+  // Reduce within the block
+  extern __shared__ float smem[];
+  smem[threadIdx.x] = r;
+
+  // First warp will perform reductions across warps
+  __syncthreads();
+  if ((threadIdx.x / 32) == 0) {
+    r = init;
+    for (int i = 0; i < blockDim.x; i += 32) {
+      r = reduceOp(r, modifyOp(smem[i + threadIdx.x]));
+    }
+
+    // Each lane participating writes out a value
+    smem[threadIdx.x] = r;
+  }
+
+  // First thread will perform reductions across the block
+  __syncthreads();
+  if (threadIdx.x == 0) {
+    r = init;
+#pragma unroll
+    for (int i = 0; i < 32; ++i) {
+      r = reduceOp(r, modifyOp(smem[i]));
+    }
```

Comment:
Sounds good re: __syncthreads().

FYI, removing the __syncthreads() for warp synchronous programming (admittedly sketchy based on the wording above) takes timing from 129.85 us -> 85.214 us here.
Being safe again, using intra-warp shuffles (CC 3.0+) for the bottom level takes timing down to 85.214 us (threads write to smem as usual; first warp loads from smem with a warp-wide stride and then reduces via shuffles). Using intra-warp shuffles for two-level reduction (reduce within each warp using shuffles; lane 0 in each warp writes out value to SM; first warp loads those values and reduces within itself using shuffles) takes timing down to 75.870 us with a lot less smem, which will have occupancy benefits.

```
394.63ms  2.8160us              (1 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [352]
394.65ms  2.8800us              (4 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [354]
394.66ms  2.9440us             (16 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [356]
394.68ms  4.5440us             (64 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [358]
394.69ms  10.080us            (256 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [360]
394.70ms  34.015us           (1024 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [362]
394.74ms  129.85us           (4096 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp1(void) [364]
394.87ms  2.9120us              (1 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [366]
394.87ms  2.7520us              (4 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [368]
394.88ms  2.7520us             (16 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [370]
394.89ms  4.2560us             (64 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [372]
394.89ms  9.5040us            (256 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [374]
394.91ms  28.543us           (1024 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [376]
394.95ms  85.214us           (4096 1 1)       (512 1 1)        12  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarp2(void) [378]
395.04ms  2.4000us              (1 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [380]
395.04ms  2.5280us              (4 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [382]
395.05ms  2.5600us             (16 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [384]
395.05ms  3.8720us             (64 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [386]
395.06ms  8.5110us            (256 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [388]
395.07ms  25.568us           (1024 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [390]
395.10ms  79.134us           (4096 1 1)       (512 1 1)        10  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle1(void) [392]
395.18ms  1.6960us              (1 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [394]
395.18ms  1.9520us              (4 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [396]
395.19ms  1.9520us             (16 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [398]
395.19ms  2.7200us             (64 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [400]
395.20ms  5.9520us            (256 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [402]
395.21ms  19.904us           (1024 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [404]
395.23ms  75.870us           (4096 1 1)       (512 1 1)         8       64B        0B         -           -   Tesla K40m (0)         1         7  reduceByWarpWithShuffle2(void) [406]
395.31ms  3.1360us              (1 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [408]
395.31ms  3.2000us              (4 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [410]
395.32ms  3.4240us             (16 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [412]
395.32ms  5.8240us             (64 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [414]
395.33ms  15.744us            (256 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [416]
395.35ms  55.998us           (1024 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [418]
395.41ms  218.56us           (4096 1 1)       (512 1 1)         8  2.0480KB        0B         -           -   Tesla K40m (0)         1         7  reduceTree(void) [420]
```